# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [6]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [7]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [8]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [9]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [10]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [11]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [12]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [13]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [14]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [15]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [16]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [17]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [18]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [19]:
# Write your code here
df.nunique()

Season     32
Daynum    133
Wteam     364
Wscore    130
Lteam     364
Lscore    117
Wloc        3
Numot       7
dtype: int64

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [20]:
df.value_counts

<bound method DataFrame.value_counts of         Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]>

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [21]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [22]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [23]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [24]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [25]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [26]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [27]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [28]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [29]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [30]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [31]:
# Sort the DataFrame by season
data_sorted_by_season = df.sort_values('Season')

# Sort the DataFrame by winning team
data_sorted_by_winning_team = df.sort_values('Wteam')

# Sort the DataFrame by winning score
data_sorted_by_winning_score = df.sort_values('Wscore')

# Select rows from index 100 to 200 and columns for season, winning team, and winning score
selected_data = data_sorted_by_season.iloc[100:201, [0, 1, 2]]

# Print the selected data
print(selected_data)


      Season  Daynum  Wteam
2525    1985     100   1406
2526    1985     100   1410
2527    1985     100   1423
2528    1985     100   1429
2529    1985     100   1440
...      ...     ...    ...
2447    1985      98   1212
2448    1985      98   1216
2449    1985      98   1233
2450    1985      98   1249
2425    1985      96   1449

[101 rows x 3 columns]


**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [32]:
# Write your code here
# Find the game with the highest winning score from the three subsets
max_winning_score = max(
    data_sorted_by_season['Wscore'].max(),
    data_sorted_by_winning_team['Wscore'].max(),
    data_sorted_by_winning_score['Wscore'].max()
)

# Find the corresponding season and winning team for the game with the highest winning score
game_with_highest_score = df[df['Wscore'] == max_winning_score]
season = game_with_highest_score['Season'].values[0]
winning_team = game_with_highest_score['Wteam'].values[0]

# Print the season and winning team for the game with the highest winning score
print("Season:", season)
print("Winning Team:", winning_team)


Season: 1991
Winning Team: 1258


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [33]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [34]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [35]:
# Create the 'ScoreDifference' column
df['ScoreDiff'] = abs(df['Wscore'] - df['Lscore'])

# Calculate the average 'ScoreDifference' for all games
average_score_difference = df['ScoreDiff'].mean()

# Filter the DataFrame to include only games with 'ScoreDifference' greater than the average
filtered_data = df[df['ScoreDiff'] > average_score_difference]


# Print the filtered DataFrame
print(filtered_data)


        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDiff
0         1985      20   1228      81   1328      64    N      0         17
3         1985      25   1165      70   1432      54    H      0         16
6         1985      25   1228      64   1226      44    N      0         20
8         1985      25   1260      98   1133      80    H      0         18
10        1985      25   1307     103   1288      71    H      0         32
...        ...     ...    ...     ...    ...     ...  ...    ...        ...
145280    2016     131   1401      71   1261      38    N      0         33
145282    2016     131   1433      76   1172      54    N      0         22
145284    2016     132   1114      70   1419      50    N      0         20
145285    2016     132   1163      72   1272      58    N      0         14
145288    2016     132   1386      87   1433      74    N      0         13

[57227 rows x 9 columns]


**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [36]:
# Find the row with the highest 'ScoreDifference'
max_score_diff_row = filtered_data.loc[filtered_data['ScoreDiff'].idxmax()]

# Retrieve the corresponding season, winning team, and losing team
season = max_score_diff_row['Season']
winning_team = max_score_diff_row['Wteam']
losing_team = max_score_diff_row['Lteam']

# Print the results
print("Season:", season)
print("Winning Team:", winning_team)
print("Losing Team:", losing_team)



Season: 1996
Winning Team: 1409
Losing Team: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [37]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [38]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [39]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [40]:
df.values[0][0]

1985

*Q**: Group the DataFrame by season and find the average winning score for each season.

In [41]:
# Group the DataFrame by season and calculate the mean of the 'Wscore' column
average_winning_score = df.groupby('Season')['Wscore'].mean()

# Print the average winning score for each season
print(average_winning_score.to_string())


Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [42]:
# Group the DataFrame by winning team and calculate the maximum of the 'Wscore' column
max_winning_score = df.groupby('Wteam')['Wscore'].max()

# Print the maximum winning score for each team
print(max_winning_score.to_string())


Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
1106    151
1107    108
1108    120
1109    137
1110    128
1111    111
1112    133
1113    123
1114    125
1115    106
1116    166
1117    116
1118     80
1119    104
1120    120
1121    103
1122    115
1123    115
1124    116
1125    107
1126    112
1127     94
1128     92
1129    116
1130    123
1131    110
1132    136
1133    139
1134    101
1135    115
1136    103
1137    116
1138    110
1139    144
1140    128
1141    115
1142    122
1143    121
1144    104
1145    108
1146    112
1147    124
1148    129
1149    114
1150    117
1151    125
1152    128
1153    116
1154    112
1155    120
1156    121
1157    115
1158    116
1159    111
1160    132
1161    108
1162    108
1163    130
1164    110
1165    108
1166    111
1167     98
1168    117
1169    123
1170    110
1171    105
1172    110
1173    118
1174    113
1175    124
1176    103
1177    125
1178    115
1179    109
1180    110
1181    130
1182    129
1183    10

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [43]:
# Group the DataFrame by both season and winning team
season_winning_team_group = df.groupby(['Season', 'Wteam'])

# Calculate the mean winning score for each group
average_winning_score = season_winning_team_group['Wscore'].mean()

# Find the team with the highest average winning score for each season
team_with_highest_score = average_winning_score.groupby('Season').idxmax().apply(lambda x: x[1])

# Create a new DataFrame to store the results
result_df = pd.DataFrame({'Season': team_with_highest_score.index, 'WinningTeam': team_with_highest_score.values})

# Print the result
print(result_df)


    Season  WinningTeam
0     1985         1328
1     1986         1109
2     1987         1380
3     1988         1258
4     1989         1258
5     1990         1258
6     1991         1380
7     1992         1380
8     1993         1380
9     1994         1380
10    1995         1206
11    1996         1206
12    1997         1254
13    1998         1395
14    1999         1317
15    2000         1395
16    2001         1395
17    2002         1242
18    2003         1395
19    2004         1194
20    2005         1311
21    2006         1190
22    2007         1440
23    2008         1440
24    2009         1377
25    2010         1440
26    2011         1449
27    2012         1370
28    2013         1377
29    2014         1322
30    2015         1322
31    2016         1146


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [62]:
# Group the DataFrame by both season and winning team
season_winning_team_group = df.groupby(['Season', 'Wteam'])

# Count the number of wins for each team in each season
wins_count = season_winning_team_group.size()

# Reset the index of the wins_count DataFrame
wins_count = wins_count.reset_index()

# Rename the columns of the wins_count DataFrame
wins_count.columns = ['Season', 'TeamNumber', 'WinsCount']

# Print the wins_count DataFrame
print(wins_count)


       Season  TeamNumber  WinsCount
0        1985        1102          5
1        1985        1103          9
2        1985        1104         21
3        1985        1106         10
4        1985        1108         19
...       ...         ...        ...
10167    2016        1460         20
10168    2016        1461         12
10169    2016        1462         27
10170    2016        1463         21
10171    2016        1464          9

[10172 rows x 3 columns]


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [64]:
# Create an empty DataFrame to store the results
result_df = pd.DataFrame(columns=['Season', 'TeamNumber', 'WinsCount'])

# Find the team with the most wins for each season
team_with_most_wins = wins_count.groupby('Season')['WinsCount'].idxmax().apply(lambda x: wins_count.loc[x]['TeamNumber'])

# Populate the result_df DataFrame with the results
for season, team_number in team_with_most_wins.items():
    wins_count_for_season = wins_count[(wins_count['Season'] == season) & (wins_count['TeamNumber'] == team_number)]
    wins_count_value = wins_count_for_season['WinsCount'].values[0]
    result_df = pd.concat([result_df, pd.DataFrame({'Season': [season], 'TeamNumber': [team_number], 'WinsCount': [wins_count_value]})], ignore_index=True)

# Print the result
print(result_df)


   Season TeamNumber WinsCount
0    1985       1385        27
1    1986       1181        32
2    1987       1424        33
3    1988       1112        31
4    1989       1328        28
5    1990       1247        29
6    1991       1116        30
7    1992       1181        28
8    1993       1231        28
9    1994       1163        27
10   1995       1116        26
11   1996       1269        31
12   1997       1242        31
13   1998       1242        33
14   1999       1181        32
15   2000       1409        29
16   2001       1181        29
17   2002       1153        30
18   2003       1166        29
19   2004       1390        29
20   2005       1228        32
21   2006       1181        30
22   2007       1242        30
23   2008       1272        33
24   2009       1272        31
25   2010       1242        32
26   2011       1242        32
27   2012       1246        32
28   2013       1211        30
29   2014       1455        33
30   2015       1246        34
31   201

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [65]:
# Group the DataFrame by the losing team
losing_team_group = df.groupby('Lteam')

# Calculate the average losing score for each team
average_losing_score = losing_team_group['Lscore'].mean()

# Print the average losing score for each team
print(average_losing_score)

# Group the DataFrame by the winning team
winning_team_group = df.groupby('Wteam')

# Calculate the average winning score for each team
average_winning_score = winning_team_group['Wscore'].mean()

# Print the average winning score for each team
print(average_winning_score)


Lteam
1101    60.586207
1102    59.201507
1103    64.117347
1104    64.374317
1105    61.675373
          ...    
1460    63.082915
1461    64.661972
1462    68.216117
1463    60.829213
1464    65.187063
Name: Lscore, Length: 364, dtype: float64
Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
          ...    
1460    75.531469
1461    75.170082
1462    79.906021
1463    71.720102
1464    73.926056
Name: Wscore, Length: 364, dtype: float64


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [68]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDiff            17
HighScoringGame      No
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDiff             7
HighScoringGame      No
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [69]:
# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Check if the winning score is greater than 100
    if row['Wscore'] > 100:
        # Set the 'HighScoringGame' column value to 'Yes'
        df.at[index, 'HighScoringGame'] = 'Yes'
    else:
        # Set the 'HighScoringGame' column value to 'No'
        df.at[index, 'HighScoringGame'] = 'No'


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [49]:
# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Check if the winning score is greater than 100
    if row['Wscore'] > 100:
        df.at[index, 'HighScoringGame'] = 'Yes'
    else:
        df.at[index, 'HighScoringGame'] = 'No'

# Print the updated DataFrame
print(df)


        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDiff  \
0         1985      20   1228      81   1328      64    N      0         17   
1         1985      25   1106      77   1354      70    H      0          7   
2         1985      25   1112      63   1223      56    H      0          7   
3         1985      25   1165      70   1432      54    H      0         16   
4         1985      25   1192      86   1447      74    H      0         12   
...        ...     ...    ...     ...    ...     ...  ...    ...        ...   
145284    2016     132   1114      70   1419      50    N      0         20   
145285    2016     132   1163      72   1272      58    N      0         14   
145286    2016     132   1246      82   1401      77    N      1          5   
145287    2016     132   1277      66   1345      62    N      0          4   
145288    2016     132   1386      87   1433      74    N      0         13   

       HighScoringGame  
0                   No  
1

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [50]:
# Create a dictionary to store the highest score difference for each season
highest_score_diff = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    season = row['Season']
    score_diff = row['Wscore'] - row['Lscore']
    
    # Check if the current score difference is higher than the previous highest for the season
    if season in highest_score_diff:
        if score_diff > highest_score_diff[season]['ScoreDiff']:
            highest_score_diff[season] = {'Game': index, 'ScoreDiff': score_diff}
    else:
        highest_score_diff[season] = {'Game': index, 'ScoreDiff': score_diff}

# Print the game with the highest score difference for each season
for season, info in highest_score_diff.items():
    game = info['Game']
    score_diff = info['ScoreDiff']
    print(f"Season: {season}, Game: {game}, Score Difference: {score_diff}")


Season: 1985, Game: 236, Score Difference: 60
Season: 1986, Game: 4731, Score Difference: 84
Season: 1987, Game: 8240, Score Difference: 73
Season: 1988, Game: 12046, Score Difference: 68
Season: 1989, Game: 16677, Score Difference: 70
Season: 1990, Game: 19502, Score Difference: 76
Season: 1991, Game: 25161, Score Difference: 68
Season: 1992, Game: 27997, Score Difference: 82
Season: 1993, Game: 33858, Score Difference: 81
Season: 1994, Game: 36404, Score Difference: 69
Season: 1995, Game: 39858, Score Difference: 74
Season: 1996, Game: 44653, Score Difference: 91
Season: 1997, Game: 49033, Score Difference: 80
Season: 1998, Game: 52600, Score Difference: 66
Season: 1999, Game: 57618, Score Difference: 75
Season: 2000, Game: 60616, Score Difference: 75
Season: 2001, Game: 65999, Score Difference: 75
Season: 2002, Game: 70737, Score Difference: 59
Season: 2003, Game: 74791, Score Difference: 72
Season: 2004, Game: 79181, Score Difference: 70
Season: 2005, Game: 83372, Score Difference:

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [51]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [52]:
# Calculate the total number of games played by each team
win_counts = df['Wteam'].value_counts()
loss_counts = df['Lteam'].value_counts()
total_counts = win_counts.add(loss_counts, fill_value=0)

# Print the total number of games played by each team
print(total_counts)


1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Length: 364, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [53]:
# Create the 'ScoreDifference' column using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Use groupby() and idxmax() to find the game with the highest score difference for each season
max_score_diff_per_season = df.groupby('Season')['ScoreDifference'].idxmax()

# Retrieve the corresponding rows from the original DataFrame
games_with_max_score_diff = df.loc[max_score_diff_per_season]

# Print the games with the highest score difference for each season
print(games_with_max_score_diff[['Season', 'Wteam', 'Lteam', 'ScoreDifference']])


        Season  Wteam  Lteam  ScoreDifference
236       1985   1361   1288               60
4731      1986   1314   1264               84
8240      1987   1155   1118               73
12046     1988   1328   1147               68
16677     1989   1242   1135               70
19502     1990   1181   1217               76
25161     1991   1163   1148               68
27997     1992   1116   1126               82
33858     1993   1328   1197               81
36404     1994   1228   1152               69
39858     1995   1246   1404               74
44653     1996   1409   1341               91
49033     1997   1278   1106               80
52600     1998   1395   1410               66
57618     1999   1268   1317               75
60616     2000   1261   1212               75
65999     2001   1196   1197               75
70737     2002   1328   1183               59
74791     2003   1403   1311               72
79181     2004   1328   1115               70
83372     2005   1403   1421      

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [54]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [55]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [56]:
type(df['Wscore'])

pandas.core.series.Series

In [57]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [58]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDiff  \
0    1985      20   1228      81   1328      64    N      0         17   
1    1985      25   1106      77   1354      70    H      0          7   
2    1985      25   1112      63   1223      56    H      0          7   

  HighScoringGame  ScoreDifference  
0              No               17  
1              No                7  
2              No                7

Here's an equivalent using iloc

In [59]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDiff  \
0    1985      20   1228      81   1328      64    N      0         17   
1    1985      25   1106      77   1354      70    H      0          7   
2    1985      25   1112      63   1223      56    H      0          7   

  HighScoringGame  ScoreDifference  
0              No               17  
1              No                7  
2              No                7

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [60]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDiff          0
HighScoringGame    0
ScoreDifference    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y